Shell - using shell_fprintf from multiple contexts

nordic-krch opened the pull request:
This patch allows to use shell_fprintf from other contexts than command handler. However, it requires command to indicate that shell should keep command context in foreground (it is done by command calling `shell_command_enter()`). Shell on that error code will not print prompt and will not accept any input from the user until command is terminated with CTRL+C or `shell_command_exit()` call.

It can be used for `bt scan on` command to start scanning and get devices printed on shell until it is terminated. Note that scanning is still enabled only context is returned to shell, allowing user to type new commands (eg `scan off`).

I see following next steps:
- add macro for creating foreground command with termination handler called on termination sequence (e.g. `SHELL_FG_CMD(scan, ...., cmd_scan_on, cmd_scan_off) where scan_off function would be called on termination).
- further in the future provide console interface through shell that could be passed to the command. Command then could use input and output through shell while staying in foreground. That could be used for starting another shell-type application like python.

Additional commits:
- Added foreground example command to shell sample.
- removed checkpatch errors from shell.c

#11604 will use that for scanning so it needs to go first.

nordic-krch committed:
shell: Allow calling shell_fprintf from various contexts

Extended shell to allow command to indicate that shell should
halt not accepting any input until termination sequence is
received (CTRL+C) or shell_command_exit() is called. While shell
is in that state it is allowed to print to shell from any thread
context.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
shell: Fix code style errors in shell.c

Fixed errors and warnings reported by checkpatch.pl

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
hang on, will rebase.

Vudentz commented:
I guess -EINPROGRESS is probably a better return to indicate the command is asynchronous and I suggest adding a function to flag the execution complete.

The ctrl+c is probably a great idea to interrupt it, but it may be useful to put the command in background perhaps by introducing ctrl+z bg and fg, obviously this will make the shell a lot more complex.

nordic-krch commented:
@Vudentz 
>  -EINPROGRESS is probably a better

Good point.

> introducing ctrl+z

For sure we will get there 😄

Meanwhile, I've added `shell_command_exit` call that will be used in `bt init` in `bt_ready` callback.

nordic-krch committed:
samples: subsys: shell: Add foreground command example

Exteneded shell sample with example how to use shell_fprintf from
any context with 'foreground' command.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

jakub-uC reviewed (changes requested):
Some minor drawbacks.

nordic-krch commented:
@jarz-nordic i've applied your comments.

nordic-krch commented:
Hold it for now. I've noticed that keeping command in foreground context by returning -EINPROGRESS is not an option because context may get  switched before command completes. I will add explicit call `shell_command_enter`.

nordic-krch commented:
Added explicit call `shell_command_enter` to be called to enter foreground. Note that shell will keep command in foreground only if command return 0.

dbkinder commented on doc/subsystems/shell/shell.rst:
```diff
@@ -235,13 +235,39 @@ Simple command handler implementation:
-	might lead to incorrect text display on the screen. If any text should
-	be displayed outside of the command context, then use the
-	:ref:`logger` or `printk` function.
+:cpp:func:`shell_fprintf` or shell print macros: :c:macro:`shell_print`,
```
PR #11604 is making the same changes to this shell.rst file?  (I've commented on that PR).

nordic-krch replied:
applied you suggestions from #11604.

dbkinder reviewed (changes requested):
Doc changes here are also in PR #11604 ?

jarz-nordic commented on subsys/shell/shell.c:
```diff
@@ -1193,6 +1216,35 @@ static int shell_instance_init(const struct shell *shell, const void *p_config,
+	k_thread_abort(k_current_get());
+}
+
+static void command_exit_handler(const struct shell *shell)
```
this can be `inline` function.

jarz-nordic commented on subsys/shell/shell.c:
```diff
@@ -1193,6 +1216,35 @@ static int shell_instance_init(const struct shell *shell, const void *p_config,
+	}
+}
+
+static void kill_handler(const struct shell *shell)
```
this can be inline function.

jarz-nordic commented:
@nashif or @carlescufi : May you please merge it? I need this to finish SMP example.
